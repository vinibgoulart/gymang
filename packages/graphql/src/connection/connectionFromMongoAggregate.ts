import apm from 'elastic-apm-node';
import type { ConnectionArguments } from 'graphql-relay';
import type { Aggregate } from 'mongoose';
import type mongoose from 'mongoose';

import {
  calculateOffsets,
  getPageInfo,
  offsetToCursor,
} from './connectionFromMongoCursor';
import { timeSpan } from '../perf/timeSpan';

type ObjectId = mongoose.Schema.Types.ObjectId;

const cloneAggregate = (aggregate: Aggregate<any>): Aggregate<any> =>
  aggregate._model.aggregate(aggregate.pipeline()).allowDiskUse(true);

export type ConnectionOptionsAggregate<LoaderResult, Ctx> = {
  aggregate: Aggregate<any>;
  context: Ctx;
  args: ConnectionArguments;
  loader: (
    ctx: Ctx,
    id: string | ObjectId | Record<string, any>,
    bypassViewerCanSee?: boolean,
  ) => LoaderResult;
  raw?: boolean; // loader should receive raw result
  allowDiskUse?: boolean;
  shouldCount?: boolean;
  bypassViewerCanSee?: boolean;
};

export const getTotalCount = async (
  aggregate: Aggregate<any>,
  allowDiskUse = true,
) => {
  const resultCount: { total: number }[] = await cloneAggregate(aggregate)
    .allowDiskUse(allowDiskUse)
    .count('total');

  return resultCount.length ? resultCount[0].total : 0;
};

/**
 * Your aggregate must return documents with _id fields
 *  those _id's are the ones going to be passed to the loader function
 */
export const connectionFromMongoAggregate = async <LoaderResult, Ctx>({
  aggregate,
  context,
  args = {},
  loader,
  raw = false,
  allowDiskUse = true,
  shouldCount = false,
  bypassViewerCanSee = false,
}: ConnectionOptionsAggregate<LoaderResult, Ctx>) => {
  // https://github.com/Automattic/mongoose/blob/367261e6c83e7e367cf0d3fbd2edea4c64bf1ee2/lib/aggregate.js#L46
  const clonedAggregate = cloneAggregate(aggregate).allowDiskUse(allowDiskUse);

  let totalCount = null;

  if (shouldCount) {
    const timespanKey = `connectionFromMongoAggregate-totalCount-${aggregate._model.collection.collectionName}`;
    const span = apm.startSpan(timespanKey);

    const end = timeSpan();

    totalCount = await getTotalCount(aggregate, allowDiskUse);

    if (span) {
      span.end();
    }

    if (process.env.DEBUG_PERF === 'true') {
      // eslint-disable-next-line
      console.log(`${timespanKey}:${end()}`);
    }
  }

  const {
    first,
    last,
    before,
    after,
    skip,
    limit,
    beforeOffset,
    afterOffset,
    startOffset,
    endOffset,
    startCursorOffset,
    endCursorOffset,
  } = calculateOffsets({ args, totalCount });

  // If supplied slice is too large, trim it down before mapping over it.
  if (skip > 0) {
    clonedAggregate.skip(skip);
  }

  // limit should never be 0 because $slice returns an error if it is.
  clonedAggregate.limit(limit || 1);

  const timespanKey = `connectionFromMongoAggregate-${aggregate._model.collection.collectionName}`;
  const span = apm.startSpan(timespanKey);

  if (span) {
    span.setLabel('pipeline', JSON.stringify(clonedAggregate.pipeline()));
  }

  // avoid large objects retrieval from collection
  const slice: { _id: ObjectId }[] = await (raw
    ? clonedAggregate
    : clonedAggregate.project('_id'));

  const end = timeSpan();

  if (span) {
    span.end();
  }

  if (process.env.DEBUG_PERF === 'true') {
    // eslint-disable-next-line
    console.log(`${timespanKey}:${end()}`);
  }

  const edges: {
    cursor: string;
    node: LoaderResult;
  }[] = slice.map((value, index) => ({
    cursor: offsetToCursor(startOffset + index),
    node: loader(context, raw ? value : value._id, bypassViewerCanSee),
  }));

  return {
    edges,
    count: totalCount,
    endCursorOffset,
    startCursorOffset,
    pageInfo: getPageInfo({
      edges,
      before,
      after,
      first,
      last,
      afterOffset,
      beforeOffset,
      startOffset,
      endOffset,
      totalCount,
      startCursorOffset,
      endCursorOffset,
    }),
  };
};
