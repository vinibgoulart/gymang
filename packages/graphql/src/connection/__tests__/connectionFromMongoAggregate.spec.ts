import { cleanupTest, connectMongoose, disconnectMongoose } from '@gymang/test';

import GroupModel from '../__fixtures__/GroupFixtureModel';
import {
  createMongooseGroup,
  createMongooseUser,
} from '../__fixtures__/helpers';
import { connectionFromMongoAggregate } from '../connectionFromMongoAggregate';

beforeAll(connectMongoose);

beforeEach(cleanupTest);

afterAll(disconnectMongoose);

// failing by sort
it.skip('should return connection from mongo aggregate', async () => {
  const userA = await createMongooseUser();
  const userB = await createMongooseUser();

  await createMongooseUser();
  await createMongooseUser();

  await createMongooseGroup({ users: [userA._id, userB._id] });

  // simple aggregate to return all users inside any group
  const aggregate = GroupModel.aggregate([
    {
      $lookup: {
        from: 'UserFixture',
        localField: 'users',
        foreignField: '_id',
        as: 'users',
      },
    },
    {
      // remove empty groups
      $match: { users: { $exists: true, $ne: [] } },
    },
    {
      // promote each user to a new document
      $unwind: '$users',
    },
    {
      $sort: {
        _id: 1,
      },
    },
    {
      $replaceRoot: { newRoot: '$users' },
    },
  ]).allowDiskUse(true);

  const context = {
    // got it throwing a ðŸŽ²
    randomValue: 1,
  };

  const loader = jest.fn();

  loader.mockReturnValue('user');

  const argsFirstPage = { first: 1 };

  const resultFirstPage = await connectionFromMongoAggregate({
    aggregate,
    context,
    args: argsFirstPage,
    loader,
    shouldCount: true,
  });

  expect(loader).toHaveBeenCalledTimes(1);
  expect(loader.mock.calls[0]).toEqual([context, userA._id]);
  expect(resultFirstPage).toMatchSnapshot();

  const argsSecondPage = { after: resultFirstPage.pageInfo.endCursor };

  const resultSecondPage = await connectionFromMongoAggregate({
    aggregate,
    context,
    args: argsSecondPage,
    loader,
    shouldCount: true,
  });

  expect(loader).toHaveBeenCalledTimes(2);
  expect(loader.mock.calls[1]).toEqual([context, userB._id]);
  expect(resultSecondPage).toMatchSnapshot();
});

// failing because of order
it.skip('should work with empty args', async () => {
  const userA = await createMongooseUser();
  const userB = await createMongooseUser();

  await createMongooseGroup({ users: [userA._id, userB._id] });

  // simple aggregate to return all users inside any group
  const aggregate = GroupModel.aggregate([
    {
      $lookup: {
        from: 'UserFixture',
        localField: 'users',
        foreignField: '_id',
        as: 'users',
      },
    },
    {
      // remove empty groups
      $match: { users: { $exists: true, $ne: [] } },
    },
    {
      // promote each user to a new document
      $unwind: '$users',
    },
    {
      $sort: {
        _id: 1,
      },
    },
    {
      $replaceRoot: { newRoot: '$users' },
    },
  ]).allowDiskUse(true);

  const context = {
    // got it throwing a ðŸŽ²
    randomValue: 1,
  };

  const loader = jest.fn();

  loader.mockReturnValue('user');

  const args = {};

  const result = await connectionFromMongoAggregate({
    aggregate,
    context,
    args,
    loader,
    shouldCount: true,
  });

  expect(loader).toHaveBeenCalledTimes(2);
  expect(loader.mock.calls[0]).toEqual([context, userA._id]);
  expect(result).toMatchSnapshot();
});

it('should work with empty args and empty result', async () => {
  // simple aggregate to return all users inside any group
  const aggregate = GroupModel.aggregate([
    {
      $lookup: {
        from: 'UserFixture',
        localField: 'users',
        foreignField: '_id',
        as: 'users',
      },
    },
    {
      // remove empty groups
      $match: { users: { $exists: true, $ne: [] } },
    },
    {
      // promote each user to a new document
      $unwind: '$users',
    },
    {
      $sort: {
        _id: 1,
      },
    },
    {
      $replaceRoot: { newRoot: '$users' },
    },
  ]).allowDiskUse(true);

  const context = {
    // got it throwing a ðŸŽ²
    randomValue: 1,
  };

  const loader = jest.fn();

  loader.mockReturnValue('user');

  const args = {};

  const result = await connectionFromMongoAggregate({
    aggregate,
    context,
    args,
    loader,
    shouldCount: true,
  });

  expect(loader).not.toHaveBeenCalled();
  expect(result).toMatchSnapshot();
});

// sort failing
it.skip('should return connection from mongo aggregate with raw', async () => {
  const userA = await createMongooseUser();
  const userB = await createMongooseUser();

  await createMongooseUser();
  await createMongooseUser();

  await createMongooseGroup({ users: [userA._id, userB._id] });

  // simple aggregate to return all users inside any group
  const aggregate = GroupModel.aggregate([
    {
      $lookup: {
        from: 'UserFixture',
        localField: 'users',
        foreignField: '_id',
        as: 'users',
      },
    },
    {
      // remove empty groups
      $match: { users: { $exists: true, $ne: [] } },
    },
    {
      // promote each user to a new document
      $unwind: '$users',
    },
    {
      $sort: {
        _id: 1,
      },
    },
    {
      $replaceRoot: { newRoot: '$users' },
    },
  ]).allowDiskUse(true);

  const context = {
    // got it throwing a ðŸŽ²
    randomValue: 1,
  };

  const loader = jest.fn();

  loader.mockReturnValue('user');

  const argsFirstPage = { first: 1 };

  const resultFirstPage = await connectionFromMongoAggregate({
    aggregate,
    context,
    args: argsFirstPage,
    loader,
    raw: true,
    shouldCount: true,
  });

  expect(loader).toHaveBeenCalledTimes(1);
  expect(loader.mock.calls[0][1].name).toEqual(userA.name);
  expect(resultFirstPage).toMatchSnapshot();
});

it('should not return negative limit', async () => {
  const userA = await createMongooseUser();
  const userB = await createMongooseUser();

  await createMongooseUser();
  await createMongooseUser();

  await createMongooseGroup({ users: [userA._id, userB._id] });

  // simple aggregate to return all users inside any group
  const aggregate = GroupModel.aggregate([
    {
      $lookup: {
        from: 'UserFixture',
        localField: 'users',
        foreignField: '_id',
        as: 'users',
      },
    },
    {
      // remove empty groups
      $match: { users: { $exists: true, $ne: [] } },
    },
    {
      // promote each user to a new document
      $unwind: '$users',
    },
    {
      $sort: {
        _id: 1,
      },
    },
    {
      $replaceRoot: { newRoot: '$users' },
    },
  ]).allowDiskUse(true);

  const context = {
    // got it throwing a ðŸŽ²
    randomValue: 1,
  };

  const loader = jest.fn();

  loader.mockReturnValue('user');

  const args = { first: 10, after: 'bW9uZ286OQ==' };

  const result = await connectionFromMongoAggregate({
    aggregate,
    context,
    args,
    loader,
    shouldCount: true,
  });

  expect(result).toMatchSnapshot();
});

it('should return correct connection and hasNextPage when using shouldCount=false', async () => {
  const userA = await createMongooseUser();
  const userB = await createMongooseUser();
  const userC = await createMongooseUser();
  const userD = await createMongooseUser();
  const userE = await createMongooseUser();
  const userF = await createMongooseUser();

  await createMongooseUser();
  await createMongooseUser();

  await createMongooseGroup({
    users: [userA._id, userB._id, userC._id, userD._id, userE._id, userF._id],
  });

  // simple aggregate to return all users inside any group
  const aggregate = GroupModel.aggregate([
    {
      $lookup: {
        from: 'UserFixture',
        localField: 'users',
        foreignField: '_id',
        as: 'users',
      },
    },
    {
      // remove empty groups
      $match: { users: { $exists: true, $ne: [] } },
    },
    {
      // promote each user to a new document
      $unwind: '$users',
    },
    {
      $sort: {
        _id: 1,
      },
    },
    {
      $replaceRoot: { newRoot: '$users' },
    },
  ]).allowDiskUse(true);

  const context = {
    // got it throwing a ðŸŽ²
    randomValue: 1,
  };

  const loader = jest.fn();

  loader.mockReturnValue('user');

  const args = {
    first: 5,
  };

  const result = await connectionFromMongoAggregate({
    aggregate,
    context,
    args,
    loader,
    shouldCount: false,
  });

  expect(loader).toHaveBeenCalledTimes(5);
  expect(result.edges.length).toBe(5);
  expect(result).toMatchSnapshot();
});
