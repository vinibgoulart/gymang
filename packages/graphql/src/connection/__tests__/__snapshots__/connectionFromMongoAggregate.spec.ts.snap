// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not return negative limit 1`] = `
{
  "count": 2,
  "edges": [],
  "endCursorOffset": 10,
  "pageInfo": {
    "endCursor": null,
    "hasNextPage": false,
    "hasPreviousPage": true,
    "startCursor": null,
  },
  "startCursorOffset": 10,
}
`;

exports[`should return connection from mongo aggregate 1`] = `
{
  "count": 2,
  "edges": [
    {
      "cursor": "bW9uZ286MA==",
      "node": "user",
    },
  ],
  "endCursorOffset": 1,
  "pageInfo": {
    "endCursor": "bW9uZ286MA==",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "bW9uZ286MA==",
  },
  "startCursorOffset": 0,
}
`;

exports[`should return connection from mongo aggregate 2`] = `
{
  "count": 2,
  "edges": [
    {
      "cursor": "bW9uZ286MQ==",
      "node": "user",
    },
  ],
  "endCursorOffset": 2,
  "pageInfo": {
    "endCursor": "bW9uZ286MQ==",
    "hasNextPage": false,
    "hasPreviousPage": true,
    "startCursor": "bW9uZ286MQ==",
  },
  "startCursorOffset": 1,
}
`;

exports[`should return connection from mongo aggregate with raw 1`] = `
{
  "count": 2,
  "edges": [
    {
      "cursor": "bW9uZ286MA==",
      "node": "user",
    },
  ],
  "endCursorOffset": 1,
  "pageInfo": {
    "endCursor": "bW9uZ286MA==",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "bW9uZ286MA==",
  },
  "startCursorOffset": 0,
}
`;

exports[`should return correct connection and hasNextPage when using shouldCount=false 1`] = `
{
  "count": null,
  "edges": [
    {
      "cursor": "bW9uZ286MA==",
      "node": "user",
    },
    {
      "cursor": "bW9uZ286MQ==",
      "node": "user",
    },
    {
      "cursor": "bW9uZ286Mg==",
      "node": "user",
    },
    {
      "cursor": "bW9uZ286Mw==",
      "node": "user",
    },
    {
      "cursor": "bW9uZ286NA==",
      "node": "user",
    },
  ],
  "endCursorOffset": 5,
  "pageInfo": {
    "endCursor": "bW9uZ286NA==",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "bW9uZ286MA==",
  },
  "startCursorOffset": 0,
}
`;

exports[`should work with empty args 1`] = `
{
  "count": 2,
  "edges": [
    {
      "cursor": "bW9uZ286MA==",
      "node": "user",
    },
    {
      "cursor": "bW9uZ286MQ==",
      "node": "user",
    },
  ],
  "endCursorOffset": 2,
  "pageInfo": {
    "endCursor": "bW9uZ286MQ==",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "bW9uZ286MA==",
  },
  "startCursorOffset": 0,
}
`;

exports[`should work with empty args and empty result 1`] = `
{
  "count": 0,
  "edges": [],
  "endCursorOffset": 0,
  "pageInfo": {
    "endCursor": null,
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": null,
  },
  "startCursorOffset": 0,
}
`;
